SU2_CFD/include/variables/CVariable.hpp

```diff
@@ -94,6 +94,101 @@ class CVariable {
   su2matrix<int> AD_InputIndex;    /*!< \brief Indices of Solution variables in the adjoint vector. */
   su2matrix<int> AD_OutputIndex;   /*!< \brief Indices of Solution variables in the adjoint vector after having been updated. */
 
+  VectorType SolutionExtra; // Stores the full adjoint solution
+  VectorType ExternalExtra; // External storage for the adjoint value (i.e. for the OF mainly)
+
+  VectorType SolutionExtra_BGS_k; // Intermediate storage, enables cross term extraction as that is also pushed to Solution
+
+  su2vector<int> AD_InputIndex_Extra; // Stores input index of the Extra vars
+  su2vector<int> AD_OutputIndex_Extra; // Stores output index of the Extra vars
+
+  VectorType SolutionExtra_Direct; // Stores values of the original direct solution, for later resetting
+  su2vector<su2double*> SolutionExtra_OriginAdresses; // Stores pointers to the original vars, these vars are resetted via this pointer vector
+
+ public: // Make public for easy debug access
+
+  // Add an su2double to the origin addresses for the extra solution variables
+  // only called by direct solver
+  inline void AddSolutionExtra_OriginAdress(su2double* val) {
+    SolutionExtra_OriginAdresses.resize(1); // here a push_back would be premium
```

Comment:
Just give it an interface that takes the vector of initial values, these give you the size and the direct values for you to store and then callback the direct solver to do the reset.

Reply:
For now I made the (small & convenient) change to std::vector  here: [](https://github.com/su2code/SU2/pull/1536/commits/fab7a4d42e9e66967c9013cd1c0434e8af15f4a5)[fab7a4d](https://github.com/su2code/SU2/pull/1536/commits/fab7a4d42e9e66967c9013cd1c0434e8af15f4a5)